metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -222,81 +222,93 @@ def standalone_events(self) -> Iterable[AuditEvent]:
-        """
-        return self.read_events.sql_query_iterator(query)  # type: ignore
-
-    def query_count(self, timestamp: str, resource: str) -> int:
-        query = f"""
-        SELECT
-          COUNT(q.query) count
-        FROM
-          {self.read_events.tablename} r
-          LEFT JOIN {self.query_events.tablename} q ON r.name = q.key
-        WHERE r.timestamp = ? AND r.resource = ?
-        """
-        return self.read_events.sql_query(
-            query, params=(timestamp, resource), refs=[self.query_events]
-        )[0][0]
-
-    def query_freq(
-        self, timestamp: str, resource: str, top_n: int
-    ) -> List[Tuple[str, int]]:
-        query = f"""
-        SELECT
-          q.query,
-          COUNT(r.key) count
-        FROM
-          {self.read_events.tablename} r
-          LEFT JOIN {self.query_events.tablename} q ON r.name = q.key
-        WHERE r.timestamp = ? AND r.resource = ?
-        GROUP BY q.query
-        ORDER BY count DESC, q.query
-        LIMIT {top_n}
-        """
-        return self.read_events.sql_query(  # type: ignore
-            query, params=(timestamp, resource), refs=[self.query_events]
-        )
-
-    def user_frequencies(self, timestamp: str, resource: str) -> List[Tuple[str, int]]:
-        query = f"""
-        SELECT
-          r.user,
-          COUNT(r.key) count
-        FROM
-          {self.read_events.tablename} r
-        WHERE r.timestamp = ? AND r.resource = ?
-        GROUP BY r.user
-        ORDER BY count DESC, r.user
+    @staticmethod
+    def usage_statistics_query(top_n: int) -> str:
+        return f"""
+        SELECT a.timestamp, a.resource, a.query_count, b.query_freq, c.user_freq, d.column_freq FROM (
+            SELECT
+                r.timestamp,
+                r.resource,
+                COUNT(q.query) query_count
+            FROM
+                read_events r
+                LEFT JOIN query_events q ON r.name = q.key
+            GROUP BY r.timestamp, r.resource
+        ) a
+        LEFT JOIN (
+            SELECT timestamp, resource, json_group_array(json_array(query, query_freq)) as query_freq FROM (
+                SELECT
+                    r.timestamp,
+                    r.resource,
+                    q.query,
+                    COUNT(r.key) as query_freq,
+                    ROW_NUMBER() over (PARTITION BY r.timestamp, r.resource, q.query ORDER BY COUNT(r.key) DESC, q.query) as rank
+                FROM
+                    read_events r
+                    LEFT JOIN query_events q ON r.name = q.key
+                GROUP BY r.timestamp, r.resource, q.query
+                ORDER BY r.timestamp, r.resource, query_freq DESC, q.query
+            ) WHERE rank <= {top_n}
+            GROUP BY timestamp, resource
+        ) b ON a.timestamp = b.timestamp AND a.resource = b.resource
+        LEFT JOIN (
+            SELECT timestamp, resource, json_group_array(json_array(user, user_freq)) as user_freq FROM (
```

Comment:
wow these queries ended up way more complex than I was expecting

Reply:
Was a bit simpler when I was doing `UNION ALL`, but this allows for much simpler python code